Disallow writes that clobber an existing timestamp.

bdarnell opened the pull request:
This could happen in tests that use manual clocks, and produces
extremely hard-to-debug failures.

Restructure mvccPutInternal for clarity and to untangle the several
places that examine timestamps.

bdarnell committed:
Disallow writes that clobber an existing timestamp.

This could happen in tests that use manual clocks, and produces
extremely hard-to-debug failures.

Restructure mvccPutInternal for clarity and to untangle the several
places that examine timestamps.

tamird commented on storage/engine/mvcc.go:
```diff
@@ -653,54 +653,47 @@ func mvccPutInternal(engine Engine, ms *proto.MVCCStats, key proto.Key, timestam
+			// There is an uncommitted write intent.
+
+			if txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID) {
+				// The current Put operation does not come from the same
```
perhaps this should panic?

bdarnell replied:
Looks like this comment has rotted. This does happen and we rely on getting the WriteIntentError instead of a panic. I'll remove the second sentence.

BramGruneir commented on storage/engine/mvcc.go:
```diff
@@ -653,54 +653,47 @@ func mvccPutInternal(engine Engine, ms *proto.MVCCStats, key proto.Key, timestam
 				versionKey := mvccEncodeTimestamp(metaKey, meta.Timestamp)
 				if err := engine.Clear(versionKey); err != nil {
 					return err
 				}
```
Is there a case where none of this if statements are true?  Just for clarity, could you add a comment even if a full statement is required?

tschottdorf replied:
this can happen when the txn writes the same key twice without restarting inbetween, but in that case the meta key requires no update.

bdarnell replied:
Yes: if none of these if statements are true, then we are replacing an intent from our same transaction which was written with the same timestamp (so we don't need an explicit clear of the old value; we'll just overwrite it directly). However, this is concerning in combination with the out-of-order comments above: if a client does send two different writes to the same key with the same timestamp, we have no way of knowing which one was intended to be later if commands may get out of order. But I'm not sure if that's actually possible; I think there is higher-level logic that will prevent out-of-order RPCs from reaching this point (unless perhaps they're sent as a part of the same batch). @spencerkimball, what is the concern about reordered RPCs at this level?

vivekmenezes commented on storage/engine/mvcc.go:
```diff
@@ -653,54 +653,47 @@ func mvccPutInternal(engine Engine, ms *proto.MVCCStats, key proto.Key, timestam
+			// No pending transaction, so we can write as long as it is not
+			// in the past.
+			if !meta.Timestamp.Less(timestamp) {
+				//if timestamp.Less(meta.Timestamp) {
```
update comment

tamird replied:
remove?

tschottdorf replied:
the commented line is the reason for this PR, right?

bdarnell replied:
Removed.

tamird commented on storage/engine/mvcc_test.go:
```diff
@@ -1196,6 +1199,10 @@ func TestMVCCWriteWithDiffTimestampsAndEpochs(t *testing.T) {
 	if err := MVCCPut(engine, nil, testKey1, makeTS(0, 1), value2, txn2); err == nil {
 		t.Fatal("expected write too old error")
 	}
+	// Ties also get write too old errors.
```
s/write too old errors/`WriteTooOldError`s/g

tschottdorf commented on storage/engine/mvcc.go:
```diff
@@ -653,54 +653,47 @@ func mvccPutInternal(engine Engine, ms *proto.MVCCStats, key proto.Key, timestam
-		newMeta = &buf.newMeta
-		buf.newMeta = proto.MVCCMetadata{Txn: txn, Timestamp: timestamp}
 	}
+	newMeta := &buf.newMeta
```
could re-use `meta` here and avoid the `meta=nil` above (which probably safeguards against anyone continuing to use it)

bdarnell replied:
We need both meta and newMeta for the call to updateStatsOnPut below.

tschottdorf replied:
Not really, `meta` stays nil.

bdarnell replied:
`meta` stays nil if it was set on line 696, but if we took the other half of the `if ok` it will be non-nil. This is weird; it should probably be the other way around (`meta` only becomes non-nil if `ok` is true).

tschottdorf replied:
You're right. But why is it weird? It looks like it's set to nil precisely when there there wasn't a meta entry before, which seems like the right thing to put into `updateStatsOnPut` along with the meta entry that gets written.

bdarnell replied:
I've moved things around here so that `meta` is not set until after we've checked `if ok`.

vivekmenezes commented on storage/engine/mvcc.go:
```diff
@@ -653,54 +653,47 @@ func mvccPutInternal(engine Engine, ms *proto.MVCCStats, key proto.Key, timestam
+				// Ignore old writes from the current transaction; it's just an
+				// RPC arriving out of order.
+				return nil
+			} else if !timestamp.Equal(meta.Timestamp) {
```
Just a personal preference because you return from the other conditions above.
I prefer this else if split up with the else { followed with a // comment about everything being okay
followed by
// replace our write intent if the timestamps are unequal
if (!...) { }

}

bdarnell committed:
Move assignment of `meta` in `mvccPutInternal`.

With this change, `meta` only has a non-nil value when `engine.GetProto`
returns `ok`, instead of being set to a non-nil but bogus value and then
being cleared later.

tamird commented:
Can you write a test?

tamird commented on storage/engine/mvcc.go:
```diff
@@ -627,25 +627,24 @@ func mvccPutInternal(engine Engine, ms *proto.MVCCStats, key proto.Key, timestam
 		if value.Deleted {
 			metaKeySize, metaValSize, err = 0, 0, engine.Clear(metaKey)
 		} else {
+			meta := &proto.MVCCMetadata{}
```
this local doesn't look useful:
```go
metaKeySize, metaValSize, err = PutProto(engine, metaKey, &proto.MVCCMetadata{Value: value.Value})
```

petermattis replied:
Is there a reason to be allocating a new `proto.MVCCMetadata` here? If you want to make sure it is empty, then:

```
  meta.Reset()
  meta.Value = value.Value
```

petermattis replied:
Or: `*meta = proto.MVCCMetadata{Value: value.Value}`

petermattis replied:
Ah, nevermind. `meta` doesn't exist at this point. The old code's intention was to avoid an allocation. You could do so here by using `buf.meta`. It's just scratch space.

bdarnell replied:
There is no longer an outer `meta` variable in scope here (and if there were, it would be `nil`). I suppose we could use `buf.meta` instead to avoid an allocation.

bdarnell commented:
I added a case in mvcc_test.go for the one intentional change. Everything else is just cosmetic changes that are not expected to change any behavior, so they're tested by making sure they don't break anything else. Was there a particular thing you had in mind to test?

tamird commented on storage/engine/mvcc.go:
```diff
@@ -653,54 +652,51 @@ func mvccPutInternal(engine Engine, ms *proto.MVCCStats, key proto.Key, timestam
-		newMeta = &buf.newMeta
-		buf.newMeta = proto.MVCCMetadata{Txn: txn, Timestamp: timestamp}
 	}
+	newMeta := &buf.newMeta
```
is `newMeta` really `oldNewMeta` here?

bdarnell replied:
No, its a pointer so it will point to the value that is set on the next line. The introduction of the putBuffer left this code with some really strange organization.

tamird replied:
OK. Would you mind just reversing these lines? (or getting rid of `newMeta` altogether, I'd be OK with that too)

tamird commented:
LGTM
